File: python/quaternion.py

import numpy as np


class vector:

    def cross(v):
        v1, v2, v3 = v.squeeze()
        return np.array([
            [0, -v3, v2],
            [v3, 0, -v1],
            [-v2, v1, 0]
        ])


class quaternion:

    def from_theta(e, theta):
        n = 1 if not hasattr(theta, '__len__') else len(theta)
        q = np.append(e.reshape((3, 1)) * np.sin(theta / 2), np.cos(theta / 2))
        return q.reshape((4, n))

    def psi(q):
        qv = q[0:3]
        q4 = q[3]
        qcross = vector.cross(qv)
        a = q[3] * np.eye(3) - qcross
        b = - (qv.T).reshape((1, 3))
        return np.concatenate((a, b), axis=0)

    def xi(q):
        qv = q[0:3]
        q4 = q[3]
        qcross = vector.cross(qv)
        a = q[3] * np.eye(3) + qcross
        b = - (qv.T).reshape((1, 3))
        return np.concatenate((a, b), axis=0)

    def cross(q):
        return np.append(quaternion.psi(q), q, axis=1)

    def dot(q):
        return np.append(quaternion.xi(q), q, axis=1)

    def A(q):
        return quaternion.xi(q).T @ quaternion.psi(q)


if __name__ == "__main__":

    from rich import print

    q = quaternion

    rot = q.from_theta(np.array([0, 0, 1]), np.pi / 2)
    print(rot, rot.shape)

    rot_xi = q.xi(rot)
    print(rot_xi, rot_xi.shape)

    rot_cross = q.cross(rot)
    print(rot_cross, rot_cross.shape)

    rot_A = q.A(rot)
    print(rot_A, rot_A.shape)

File: matlab/quaternion.m
function quaternion = quaternion()
    quaternion.q = @q;
    quaternion.psi = @psi;
    quaternion.xi = @xi;
    quaternion.cross = @cross;
    quaternion.dot = @dot;
    quaternion.dq = @dq;
    quaternion.I = @I;
    quaternion.inv = @inv;
    quaternion.conj = @conj;
    quaternion.A = @A;
end

function q = q(e, t)
    q = [e * sin(t / 2); cos(t / 2)];
end

function psi = psi(q)
    vec = vector;
    v = q(1:3);
    psi = [q(4) * eye(3) - vec.cross(v); -v'];
end

function xi = xi(q)
    vec = vector;
    v = q(1:3);
    xi = [q(4) * eye(3) + vec.cross(v); -v'];
end

function cross = cross(q)
    cross = [psi(q), q];
end

function dot = dot(q)
    dot = [xi(q), q];
end

function A = A(q)
    A = xi(q)' * psi(q);
end

function dq = dq(q, q_c)
    dq = cross(q) * q_c;
end

function conj = conj(q)
    conj = [-q(1:3); q(4)];
end

function inv = inv(q)
    inv = conj(q) / (norm(q))^2;
end

function I = I()
    I = [0, 0, 0, 1]';
end

